clang/lib/Sema/SemaCodeComplete.cpp

```diff
@@ -1428,6 +1428,16 @@ void ResultBuilder::AddResult(Result R, DeclContext *CurContext,
 
   AdjustResultPriorityForDecl(R);
 
+  if (isa<FieldDecl>(R.Declaration)) {
+    // If result is a member in the context of an explicit-object member
+    // function, drop it because it must be accessed through the object
+    // parameter
+    if (auto *MethodDecl = dyn_cast<CXXMethodDecl>(CurContext);
```

Comment:
I don't think `CurContext` is the right `DeclContext` to check here, as it could be a nested context inside the method. For example:

```c++
struct A {
  int member;
  void foo(this A &self) {
    [&]() {
      // Completion should not offer `member`, but with the current patch it does.
      mem^
    }();
  }
};
```

I think you want something like `getFunctionLevelDeclContext(/*AllowLambda=*/false)`.

Also, as mentioned on Discord, we should do this check in `CodeCompleteOrdinaryName()` and save the result on the `ResultBuilder`. This is actually important not just for performance, but also to limit the effect of the change to cases where we're in `CodeCompleteOrdinaryName()`, since `ResultBuilder` is used for completions in other contexts as well.

Reply:
> Is storing it with a call to `setObjectTypeQualifiers` fine? Or is introducing a new get-set pair better?

I think it would actually make the most sense in an `else` branch of [this `if`](https://searchfox.org/llvm/rev/a293573c4e3e43f9f6279f075c3262ea5dc17086/clang/lib/Sema/SemaCodeComplete.cpp#4642), since `getCurrentThisType()` should return null in an explicit-object member function (due to [this check](https://searchfox.org/llvm/rev/a293573c4e3e43f9f6279f075c3262ea5dc17086/clang/lib/Sema/SemaExprCXX.cpp#1214)) (which suggests using a different get/set pair).

Reply:
Good catch, but I would say this is an unrelated bug:

```c++
struct A {
  int foo(this A const&);
};
void bar(const A& a) {
  a.^  // foo not offered
}
```

This scenario involves `CodeCompleteMemberReferencExpr`, so I don't think it needs to influence what we do in `CodeCompleteOrdinaryName`. Let's file it as another issue to look at later?

Reply:
> This scenario involves CodeCompleteMemberReferencExpr, so I don't think it needs to influence what we do in CodeCompleteOrdinaryName.

Sounds good!

>  Let's file it as another issue to look at later?

Looks like it's the first part of #109608.